### DISCUSSION TITLE
How can i conditionally render child livewire components, when route is pointing to the parent?

### DISCUSSION THREAD

**Discussion by joshembling** [livewire/livewire#8629] [1. Help]
I have a scenario where I can only render one uri, of which different views should be rendered by a query string value. Because of this I need to conditionally render different blade views based on `$this->mode` that comes from the parent class `FirebaseAuthForm`. 

The following code kind of works, but on the first submission, the data is cleared, full validation is not actioned (explicitly, except for rendering the first form item validation error) and the form is never submitted. 

On subsequent attempts it does actually work as expected with full validation.

It seems to me, on the first submission it is looking for some data in the parent class that it will never find, on the second submission, it calls the correct class `ResetPassword` and works as expected. 

I suppose the question is, how can I target the specific child class every time, so the user doesn't have to fill in the form twice? On this occasion it is a short password form so not 100% disastrous, but in other forms there are 10+ fields.

```php
Route::get('/members', FirebaseAuthForm::class);
```

```php
class FirebaseAuthForm extends Component
{
    #[Url]
    public $mode;

    public function render()
    {
        return view('livewire.firebase-auth-form')->with([
            'mode' => $this->mode,
        ]);
    }
}
```

```php 
// parent blade view 

@switch($mode)
    @case('resetPassword')
        <livewire:reset-password :wire:key="$mode" />
        @break
// other cases etc...
    @default
        @php
            return abort(404);
        @endphp
@endswitch
```

```php 
// Example child (using Filament form classes)

class ResetPassword extends FirebaseAuthForm implements HasForms
{
    use InteractsWithForms;

    public ?array $data = [];

    public function mount()
    {
        $this->form->fill();
    }

    public function rules()
    {
        return [
            'password' => ['confirmed', 'required'],
            'password_confirmation' => ['required', 'same:password'],
        ];
    }

    public function messages()
    {
        return [
            'content.required' => 'The :attribute is required.',
            'content.min' => 'The :attribute is too short.',
            'content.mixed_case' => 'The :attribute must contain at least one uppercase and one lowercase letter.',
            'content.numbers' => 'The :attribute must contain at least one number.',
            'content.symbols' => 'The :attribute must contain at least one symbol.',
        ];
    }

    public function form(Form $form): Form
    {
        return $form
            ->schema([
                Forms\Components\TextInput::make('password')
                    ->label('Password')
                    ->password()
                    ->revealable()
                    ->placeholder('Enter your new password')
                    ->rules(array_merge($this->rules()['password'], [new PasswordValidationRule()])),

                Forms\Components\TextInput::make('password_confirmation')
                    ->label('Confirm Password')
                    ->password()
                    ->revealable()
                    ->placeholder('Confirm your password'),
            ])
            ->statePath('data');
    }

    #[On('password-submit')]
    public function submit()
    {
        $data = $this->form->getState();

        // more business logic...

        session()->flash('message', 'Password successfully updated.');

        $this->reset();
    }

    public function render(): View
    {
        return view('livewire.member-password-reset');
    }
}
```

```php
// Child blade view

<form wire:submit.prevent="$dispatch('password-submit')" class="fi-form grid gap-y-6">
    {{ $this->form }}

// etc...
```

I should also note that targeting the child class directly in the route works as expected (see below), but as mentioned, I need to conditionally render these child Livewire components.

```php
// works
Route::get('/members', ResetPassword::class);

// ResetPassword works on second submission
Route::get('/members', FirebaseAuthForm::class);
```

**Comment by lrljoe** [comment]
Not sure what you're doing having the same URI trying to go to two different places?

```
    public function render()
    {
        return view('livewire.firebase-auth-form')->with([
            'mode' => $this->mode,
        ]);
    }
```
$mode is public, you don't need to pass it into the view, and that may well cause you some issues!

If you're wanting to share data between multiple different child components, then you need to either pass it back to the parent, pass it directly to another component or wire:model/modelable it.

If you want to pass it directly to another component, then you'll want to have all components present on the page, even if hidden, as otherwise they aren't loaded/able to be spoken to.
